[WIP] Adding Cliffords to QuantumCircuits as Operations

alexanderivrii opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

In this work-in-progress issue, we are adding a `Clifford` object to a `QuantumCircuit` as an `Operation`, without first explicitly converting this `Clifford` into a `Gate`. 

An important point is that a `Clifford` is no longer _decomposed_ as soon as it is added to a `QuantumCircuit`. The procedure `decompose_clifford` only runs when the `definition` of a `Clifford` is required, which might only be during a later transpiler pass.

To illustrate an additional benefit of adding `Clifford`s natively, there is a preliminary transpiler pass `OptimizeCliffords` that uses the ability to _compose_  `Clifford`s.

In the near future, other objects will be added as `Operation`s as well.

### Details and comments

A `Clifford` object does not inherit from `Instruction`. In order to work with `Clifford`s natively, it must support the following:
- `broadcast_arguments`: this is used for expanding (aka broadcasting) arguments. Note that there is an additional PR to improve the broadcasting code. 
- `condition`: as far as I understand, this is used for _conditional if_ instructions. I don't think it applies to `Clifford`s, but is currently required by `circuit_to_dag`. I believe this is something that we can later clean up.
- `_directive`: This is a attribute to treat like barrier for transpiler, unroller, drawer. It is currently required by `DagCircuit`. Again I think that this may be something that we can later clean up.
- `definition`: This is the function that builds the definition of the `Clifford` in terms of more basic gates (i.e. the function that essentially calls `decompose_clifford`).

alexanderivrii commented:
A few questions/comments came up in the discussion with @eliarbel and @ShellyGarion:

1. This PR undoes #7840, restoring #7087, and unfortunately leading to a slight temporary slowdown as per #7528. However, I don't think that the slowdown is significant.

2. One change, compared to #7087, is that `Instruction` now inherits from `Operation`, this makes the code a bit simpler, in that we need to check in several places `isinstance(instruction, Operation)` instead of `isinstance(instruction, [Instruction, Operation])`. I don't have a strong opinion on which alternative is better and open to suggestions.

3. The test `test_native_operations.py` needs to be reorganized (and possibly renamed). The files `optimize_cliffords.py` and `test_clifford_passes.py` are currently more for demonstration purposes (what could we do if Cliffords were added as Operations), we may or may not include these in this PR.

alexanderivrii commented on qiskit/circuit/operation.py:
```diff
@@ -59,3 +59,26 @@ def num_qubits(self):
+        """Condition for when the instruction has a conditional if."""
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def definition(self):
+        """Definition of the operation in terms of more basic gates."""
+        raise NotImplementedError
```
Currently multiple transpiler passes (`Decompose`,  `Unroll3qOrMore`, etc.) require `node.op.definition`, so this is a method that had to be implemented for `Cliffords` and that imho `Operation` should support. However, the main concern is already addressed (for `Cliffords`): the definition circuit is not constructed at the gate creation time (and only when it's actually needed). For instance, the following code:

```
qc1 = QuantumCircuit(5)
qc1.append(random_clifford(3), [4, 0, 2])
qc1.append(random_clifford(3), [4, 0, 2])
qc1.append(random_clifford(3), [4, 0, 2])
qc2 = PassManager(OptimizeCliffords()).run(qc1)
qc3 = qc2.decompose()
```

only constructs the definition circuit for a single clifford in `qc2` and only when `decompose` is called.

alexanderivrii commented on qiskit/quantum_info/operators/symplectic/clifford.py:
```diff
@@ -540,6 +548,34 @@ def _pad_with_identity(self, clifford, qargs):
+        flat_cargs = [carg for sublist in cargs for carg in sublist]
+        yield flat_qargs, flat_cargs
+
+    @property
+    def definition(self):
+        """Computes and returns the circuit for this Clifford."""
+        if self._definition is None:
+            self._definition = self.to_circuit()
+
+        return self._definition
```
> how far do we make it through a transpile call if we don't add a definition? 

I would say quite far, please see the code snippet from my answer to a previous question. 

> The long term goal is to have a transpiler pass which can synthesize Cliffords and use the transpilation context to make some intelligent choices between different synthesis methods.

I completely agree. This is also something that @ShellyGarion is interested in, as she wants to be able to choose at transpile time whether to apply a synthesis method that minimizes gate-count vs. depth and whether to apply a method that is better suited for all-to-all connectivity vs. linear-neighbor connectivity.

> can to_instruction above now return self instead of converting to a circuit and then to an instruction?

This is an interesting suggestion (that I have not tried yet). I might be wrong about this, but I am afraid that certain methods (like constructing a unitary `Operator` for a quantum circuit containing a clifford) depend on `to_operation` translating more complex objects to simpler objects.

alexanderivrii commented on qiskit/transpiler/passes/optimization/optimize_cliffords.py:
I was also thinking about this, and this should be possible, but we need to suitably permute the rows and the columns of the clifford's stabilizer tableau.

alexanderivrii commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -1261,7 +1262,11 @@ def _append(
 
-    def _update_parameter_table(self, instruction: Instruction) -> Instruction:
+    def _update_parameter_table(self, instruction: Operation) -> Operation:
+        # A generic Operation object at the moment does not require to have params.
```
I have updated the code to check `isinstance(instruction, Instruction)`. 

I don't remember the reason of not including `params` in `Operation` either. Possibly this would be a good idea, even though `Cliffords` don't use `params` to store stabilizer/destabilizer table. Do we want them to? Is it true that two `Instructions` with the same `name`, `num_qubits`, `num_clbits` and parameters are completely equivalent? Do we want this to be true?

alexanderivrii commented on qiskit/quantum_info/operators/operator.py:
```diff
@@ -75,7 +75,7 @@ def __init__(self, data, input_dims=None, output_dims=None):
             # Default initialization from list or numpy array matrix
             self._data = np.asarray(data, dtype=complex)
-        elif isinstance(data, (QuantumCircuit, Instruction)):
+        elif isinstance(data, (QuantumCircuit, Operation)):
```
Thanks! I think I have updated all the places that I could find, but maybe I've still missed a few.

alexanderivrii commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -1160,7 +1161,7 @@ def _resolve_classical_resource(self, specifier):
     def append(
         self,
-        instruction: Instruction,
+        instruction: Operation,
```
I have updated the docstring. Would it be possible to postpone renaming/deprecation to a small follow-up PR?

alexanderivrii commented on qiskit/transpiler/passes/optimization/optimize_cliffords.py:
```diff
@@ -0,0 +1,85 @@
+
+class OptimizeCliffords(TransformationPass):
+    """Combine consecutive Cliffords over the same qubits.
+    This serves as an example of extra capabilities enabled by storing
+    Cliffords natively on the circuit.
```
> I think we should look to generalize CollectMultiQBlocks or similar so that we can efficiently collect and grow blocks to resynthesize, but that's a separate issue.

Agreed. This seems like some important functionality that comes up over and over.

ikkoham reviewed (changes requested):
I understand that the path to optimize this Clifford is very important. However, I am against extending the `Clifford` in quantum_info to Operation/Instruction, etc. quantum_info and Circuit are on different levels of hierarchy in Qiskit, and confusion would create a circular dependency. For example, [Operator](https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/quantum_info/operators/operator.py) is a class that represents operators, and unitary operators are represented using this class. But, this class is not Instruction. For unitary gates, [another class](https://github.com/Qiskit/qiskit-terra/blob/48306bda792599e3bf456e2b5ea254347c386df9/qiskit/extensions/unitary.py) UnitaryGate is used for QuantumCircuit. Therefore, it is safer to create the separate class such as CliffordGate or CliffordOperation to separate the hierarchy.

alexanderivrii commented:
@ikkoham, thank you for your feedback! This PR is the result of a discussion that has been going on for quite some time, please see #5811 (and #7087), and in particular @kdk's comment https://github.com/Qiskit/qiskit-terra/issues/5811#issuecomment-920344575.

You can view this as moving in the direction of high-level-synthesis, @kdk and @jakelishman can explain this better than me.

Please note that we can already add `Cliffords` onto a `QuantumCircuit` as per:
```
qc = QuantumCircuit(4)
cliff = random_clifford(2)
qc.append(cliff, [1, 3])
```
so to an extent the confusion already exists.

From the transpilation point of view, the above currently implicitly converts `cliff` to a `Gate` (and loses the `Clifford` structure of the object).  This PR somewhat resolves this. I agree that this is not the only way to do this, and we could have (if I understand your suggestion correctly) define a class such as 

```
CliffordGate(Gate):
  def __init__(self, cliff):
    # stores Clifford object
    self._cliff = cliff
```
and then call this like
```
qc.append(CliffordGate(cliff), [1, 3])
```
but frankly I don't see much benefit (and would require us to do many changes throughout the code). 

---

We have discussed the current status of this PR during the last Terra Circuits meeting, and we have agreed that we do not want to include `definition` inside `Operation` (and we are currently discussing on how to resolve this issue). 

---

@ikkoham, to see if I understand your second point. You are suggesting a transpiler pass that would collect blocks of consecutive Clifford gates, and merge these into a single `Clifford` object, is this correct? I.e. similar to recent work on defining `LinearFunctions` (https://qiskit.org/documentation/stubs/qiskit.circuit.library.LinearFunction.html) and combining these as per 
https://qiskit.org/documentation/stable/0.35/stubs/qiskit.transpiler.passes.CollectLinearFunctions.html.

jakelishman commented:
@ikkoham: we totally agree with you - we're very much _not_ trying to add any meaningful extra code to `Clifford`, and `Clifford` shouldn't be dependent on `QuantumCircuit` at all.

The below is a _very_ rough sketch of one possible idea, but it's still under discussion and not decided at all:

We will define an `Operation` interface, which supplies only very basic information about the operation such as its number of qubits, classical bits and the types of any runtime parameters it needs.  It _doesn't_ specify a definition for itself in terms of a circuit.  This `Operation` interface can be defined at a very low level - `quantum_info` will be able to import it without importing `QuantumCircuit`, so we maintain the right hierarchy.

`QuantumCircuit` will be able to contain things that implement `Operation`, but it will be the responsibility of the transpiler to convert these into gate-level representations.  It won't be the individual classes that are responsible for this, but a consumer of them at the transpiler level.  This "higher-order" synthesis may also consider multiple `Clifford` instances at once, and may have access to layout/routing/backend information at the time it does its synthesis.  This is separating "data" (`Clifford`) from "behaviour" (various different synthesis algorithms), so we can build up everything in an extensible composable way, without needing one giant monolithic `Clifford` class (which is unmaintainable, as you say).

alexanderivrii commented:
Please refer to @jakelishman's proposal for `Operation` interface [here](https://github.com/Qiskit/rfcs/pull/25/).

Based on the recent discussions, I have removed both `definition` and `broadcast` from `Operation`'s (and in particular `Clifford`'s) API. 

The code in this PR is significantly closer to the above proposal. 

---

Regarding `definition`, we have a transpiler pass (tentatively called `HighLevelSynthesis`) that traverses the DAG and replaces each high-level-operation (only `Clifford`s for now, but in the future it will also contain `LinearFunction`s, `CNOTDihedral`s and other objects that will be native `Operation`s) by standard gates using the appropriate synthesis algorithm. 

In the future, we probably want each synthesis algorithm to be available via a plugin interface, similarly to the unitary synthesis plugin interface, thus `HighLevelSynthesis` may need to be later modified to accept a CliffordSynthesisPlugin, a LinearFunctionPlugin, etc., and possibly even the unitary synthesis plugin should also be a part of this. This is open for discussion.

As of now, I have modified the 4 preset passmanagers to run `HighLevelSynthesis` right after `UnitarySynthesis` but before `Unroll3qOrMore`, so that the `Unroll3qOrMore` pass can safely assume that every gate has a `definition` and use it for unrolling.

There is also a question whether the `Decompose` pass should decompose high-level-objects (e.g. native `Clifford`s) or leave them as they are, personally I believe that it should, so I have added `HighLevelSynthesis` to run as the first stage of `Decompose`. Comments are welcome.

---

Regarding `broadcast`. I have incorporated the ideas discussed in #7718 (and closed the latter PR). I have moved all of the broadcasting functionality to a separate `broadcast.py` file (I have not done any significant code changes, modulo replacing `return` by `yield` when appropriate), and have changed `QuantumCircuit` and `QuantumCircuitData` to call `broadcast_arguments(instruction, qargs, cargs)` instead of `instruction.broadcast_arguments(qargs, cargs)`. I have removed the functionality `broadcast_arguments` from all of the relevant places (`barrier`, `measure`, etc.) -- since they were only used by `QuantumCircuit.append`, I believe that no `deprecation` mechanism is necessary here. There seems to be a significant opportunity to further unify and clean up the code in `broadcast.py`. I have added `test_broadcast.py` that checks the functionality of all different previously supported broadcasting schemes.

jakelishman commented on qiskit/transpiler/passes/synthesis/high_level_synthesis.py:
```diff
@@ -0,0 +1,39 @@
+from qiskit.quantum_info.synthesis.clifford_decompose import decompose_clifford
+
+
+class HighLevelSynthesis(TransformationPass):
+    """Synthesize high-level objects by choosing the appropriate synthesis method based on
+    the object's name.
+    """
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        """Run the HighLevelSynthesis pass on `dag`.
+        Args:
+            dag: input dag.
+        Returns:
+            Output dag with high level objects synthesized.
+        """
+
+        for node in dag.named_nodes("clifford"):
+            decomposition = circuit_to_dag(decompose_clifford(node.op))
+            dag.substitute_node_with_dag(node, decomposition)
+
+        return dag
```
I would prefer we had more of the long-term structure set up in this pass before merge, but this is a less pressing concern to me than other parts of this PR, because it's easier for us to change transpiler-pass behaviour in subsequent releases.

ajavadia replied:
This pass seems to be a shim whose job is to call the synthesis routines (that live elsewhere) on nodes of interest. I think UnitarySynthesis is similar. And soon we will have synthesis of other higher-order operators. So I think we should consolidate them all under the same pass.

jakelishman reviewed (commented):
I accept that if we push to merge this in the 0.21 window, then it won't be feature-complete for the long term, and will only be partially supported in the near term.  It certainly wouldn't be my preference to do that, but I've reviewed under the assumption that the plan is to move ahead for an initial release in 0.21.

Even within the more permissive view of accepting an interface with partial support, there are some things I think we need to be clearer about / address before merge:
- changes to broadcasting: I don't think the changes in this PR are necessary for `Operation`, and I don't believe we settled on a long-term interface.  I would much rather we revert these changes, and simply gate the `broadcast_arguments` call in `QuantumCircuit.append` behind an `isinstance(..., Instruction)` check.  We can add broadcasting to generic `Operation`s later without breaking backwards compatibility, but we can't take it away, and my current view in favour of removing broadcasting completely from `QuantumCircuit.append` (but potentially letting the special `QuantumCircuit` methods like `measure` do it themselves if they really want).  The current change breaks backwards compatibility by removing `broadcast_arguments` from `Instruction` without warning or existing alternative, which we definitely must make safe.
- what the initial support is for `Operation`, and temporary fields (see [LINK]).  I would rather we don't put in a change to a core data structure that has only partial support (e.g. why I changed _everything_ when doing #8093), but if we're moving forwards with that, then I think we need to be hyper clear about where the support is, and preferably make errors louder and clearer when we hit unsupported parts, rather than adding kludge temporary fields that we'll have to walk back later.
- in meetings, we had discussed the fields `param_spec` and `state` as fields on `Operation`.  We can do these additively later (they both have natural "empty" values we can use as defaults), so that's ok, I just wanted to flag that the current API doesn't fully reflect what we've discussed in the past.

There's also currently merge conflicts with the `main` branch that would need solving before we can move ahead with any merge, but those shouldn't be too hard to fix.  I wrote most of #8093 with the new interface name in mind, so most of that should already have the right terminology.

jakelishman commented on qiskit/circuit/operation.py:
```diff
@@ -59,3 +59,17 @@ def num_qubits(self):
     def num_clbits(self):
         """Number of classical bits."""
         raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def _directive(self):
+        """Class attribute to treat like barrier for transpiler, unroller, drawer.
+        This is temporary, please do not use."""
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def condition(self):
+        """Condition for when the instruction has a conditional if.
+        This is temporary, please do not use."""
+        raise NotImplementedError
```
I spoke with @kdk offline about some of this as well, and while the ideal situation is that we update all getters of `condition` to do `getattr(x, 'condition', None)` in the near term and the setter of `Instruction.condition` to reject modification of non-`Instruction` conditions, if that's not possible, we could also define the scope of `Operation` support in this release to _require_ that it's converted into `Instruction` instances before any transpiler passes that might want to access its `condition` attribute.

mtreinish commented:
Given that we're tagging the release today and given the size and nature of this PR combined with this has a bunch of outstanding review comments and several merge conflicts I'm going to defer this to 0.22. Once we tag the 0.21.0rc1 later today and 0.22.0 development is opened on main we should strive to merge this as soon as possible so we can continue to iterate on it during the next 3 months.

alexanderivrii commented on qiskit/circuit/broadcast.py:
```diff
@@ -0,0 +1,222 @@
+}
+
+
+def broadcast_arguments(operation: Operation, qargs, cargs):
+    """The main function to broadcast arguments based on the operation."""
+
+    # print(f"broadcast_arguments: {operation = }, {qargs = }, {cargs = }")
+    if operation.name in broadcast_implementations.keys():
+        # Use a custom broadcast implementation when available
+        broadcaster = broadcast_implementations[operation.name]
+        yield from broadcaster(operation, qargs, cargs)
+    elif isinstance(operation, Gate):
+        # Use implementation for gates, if the operation is a gate
+        yield from broadcast_gate(operation, qargs, cargs)
+    else:
+        # Use the generic implementation otherwise
+        yield from broadcast_generic(operation, qargs, cargs)
```
@jakelishman Thanks for a very thorough review. I was unable to work for more than two weeks due to a family emergency, but I am back now. 

I agree that I shouldn't have removed `broadcast_arguments` from `Barrier` and other classes (even though these are not used anywhere in _qiskit-terra_, they are indeed part of the documented API). As per your suggestion, I will revert all the changes to broadcasting, _including the many bugs and flaws that you have pointed out_, and we will treat these in a separate PR to keep the scope of this PR slightly smaller. 

I do believe that moving all the broadcasting code in a single file is a good idea, allowing to easily see the similarities and differences between different implementations. Again, I will revert all this for now, but an alternative implementation would be to keep the file `broadcast.py` and let `broadcast_arguments` in `Barrier` be something like:

```
class Barrier(Instruction):
      def broadcast_arguments(self, qargs, cargs):
        yield from broadcast_barrier(self, qargs, cargs)
```

However, I have realized that I don't fully understand your suggestion
> just gate the `instruction.broadcast_arguments` call in `QuantumCircuit.append` behind an `isinstance(instruction, Instruction)` check to maintain the legacy behaviour, but have our default position for Operation be "Operation instances don't broadcast" right from the start. 

Do we want say `Barrier` to inherit only from `Instruction` (as in the current trunk code), from `Operation` (as in this PR), or from both? In either case, as `Instruction` inherits from `Operation`, `Barrier` will be `Operation`, but we do want to broadcast it. After thinking a bit, I presume that we want `Barrier`, etc. to inherit from `Instruction` only, and our default position be "Operation but non-Instruction instances don't broadcast". So in this PR, only two classes will directly inherit from `Operation`: `Instruction` and `Clifford`. Do you agree?

alexanderivrii commented on qiskit/transpiler/preset_passmanagers/level1.py:
```diff
@@ -252,6 +254,7 @@ def _swap_condition(property_set):
                 min_qubits=3,
                 target=target,
             ),
+            HighLevelSynthesis(),
```
Yes, I may have missed a place or two in the previous version, thanks for spotting this. I believe it should be correct to place a call to `HighLevelSynthesis` immediately after each and every call to `UnitarySynthesis`, and I agree with @ajavadia that it makes sense to combine the two (https://github.com/Qiskit/qiskit-terra/pull/7966#discussion_r912055252). This job became much easier after the code changes for `StagedPassManager`, all of my changes only belong to `preset_passmanagers/common.py`.

Ok, contrary to what I've just written, I did not add `HighLevelSynthesis` after `UnitarySynthesis` in the "optimization loop" in `level3.py`, since `Cliffords`, etc. should have been taken care prior to this. Though, this may change if/when we want to add the ability to resynthesize and to reimplement `Cliffords`. Which way do you think it's best?

alexanderivrii commented:
@jakelishman, I have fixed merge conflicts and removed the changes to broadcasting in commit [c1ab860](https://github.com/Qiskit/qiskit-terra/pull/7966/commits/c1ab8600cb4931c2678aeb69ee5518b62f1f3961). This includes restoring the functionality of  natively adding `Cliffords` to `QuantumCircuits` and making sure that no `Cliffords` remain after transpilation. I have not yet addressed any of the comments that you considered not crucial for this PR (such as for instance 
adding the fields `param_spec` and `state` on Operation). 

As per the code review, I have replaced all calls to `x._direction` by `getattr(x, "_direction", False)` in commit [aeef894](https://github.com/Qiskit/qiskit-terra/pull/7966/commits/aeef8943af73e9c042db9677cf4f63676a0e9eef). At this point, `Operation` no longer includes the method `_direction`, but does include the method `condition` (though it's no longer `abstractmethod`).

I can proceed to the comment https://github.com/Qiskit/qiskit-terra/pull/7966#discussion_r904048196, replacing getters and setters of `condition`. This will probably touch a lot of files, but should be pretty much straightforward work. Please confirm that this is indeed what you had in mind.

mtreinish commented on test/python/transpiler/test_clifford_passes.py:
```diff
@@ -0,0 +1,275 @@
+        qc.swap(1, 3)
+        qc.append(cliff2, [1, 2, 3])
+        qc.h(3)
+        # print(qc)
```
```suggestion
```

alexanderivrii replied:
thanks; fixed

mtreinish commented on test/python/transpiler/test_clifford_passes.py:
```diff
@@ -0,0 +1,275 @@
+
+        # Check that calling QuantumCircuit's decompose(), no Clifford objects remain
+        qc2 = qc.decompose()
+        # print(qc2)
```
```suggestion
```

alexanderivrii replied:
thanks; fixed

mtreinish commented on qiskit/transpiler/passes/optimization/optimize_cliffords.py:
Is this algorithm sensitive to iteration order? Using `op_nodes()` this will return a list of op nodes from the dag in `node_id` order (which is nominally insertion order except for when there are intermediate removals). This could lead to getting a weird iteration order in some cases depending on how the dag was constructed. Do you want instead to iterate over in topological order to ensure we're looking at neighboring nodes on a qubit for building the block?

```suggestion
        for node in dag.topological_op_nodes():
```

alexanderivrii replied:
This is a good suggestion, thanks! Committed this change and a small test to see this in [0d583ed](https://github.com/Qiskit/qiskit-terra/pull/7966/commits/0d583edceab35c35e63c6b1d29aa81e0f742f3f5).

mtreinish commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -1283,11 +1284,21 @@ def append(
-            instruction = CircuitInstruction(operation, qarg, carg)
-            appender(instruction)
-            instructions.add(instruction)
+        if isinstance(operation, Instruction):
+            for qarg, carg in operation.broadcast_arguments(expanded_qargs, expanded_cargs):
+                self._check_dups(qarg)
+                instruction = CircuitInstruction(operation, qarg, carg)
+                appender(instruction)
+                instructions.add(instruction)
+        else:
+            # For Operations that are non-Instructions, we use the Instruction's default method
+            for qarg, carg in Instruction.broadcast_arguments(
+                operation, expanded_qargs, expanded_cargs
+            ):
+                self._check_dups(qarg)
+                instruction = CircuitInstruction(operation, qarg, carg)
+                appender(instruction)
+                instructions.add(instruction)
```
Per @jakelishman's earlier comments do we want to implicitly use the instruction class's broadcast method here? I'm fine with this, but I feel like I'm missing a piece in my mental model of how we transition from this to a new model for broadcasting args without causing a potential compatibility issue down the road assuming we release this as is. I'm not sure what a good alternative is, since I haven't dug into the broadcasting piece in some time. I'm also probably missing something.

Is maybe doing something like:

```suggestion
        if hasattr(operation, "broadcast_arguments"):
            for qarg, carg in operation.broadcast_arguments(expanded_qargs, expanded_cargs):
                self._check_dups(qarg)
                instruction = CircuitInstruction(operation, qarg, carg)
                appender(instruction)
                instructions.add(instruction)
        else:
            # For Operations that are non-Instructions, we use the Instruction's default method
            for qarg, carg in Instruction.broadcast_arguments(
                operation, expanded_qargs, expanded_cargs
            ):
                self._check_dups(qarg)
                instruction = CircuitInstruction(operation, qarg, carg)
                appender(instruction)
                instructions.add(instruction)
```

a better choice? That way if an operation defined it's own broadcasting rules it'd work. But I might be thinking too much about this.

alexanderivrii replied:
Thanks! If I understand correctly, the decision about this during the last qiskit-dev-synthesis meeting was to leave the code as it is for now (in this stricter form), but possibly switch to what you suggest in the future.

mtreinish reviewed (commented):
Overall this LGTM, I think as a incremental first step this does a good job. Most of the PR looks like it's just fixing support for a raw operation class being passed around by making some of the `Instruction` specific checks opt-in with normal defaults along with making the `Clifford` class an operation which seems like a good next step towards implementing everything from Qiskit/rfcs#25.

I left a few inline comments but the big pieces missing for me are primarily the release notes are missing a couple things basically they just need a bit more detailed of a user facing documentation about the new features and how they could be used. I also think the higher level synthesis pass could use a couple comments explaining how it will be expanded down the road. Having a plugin interface is definitely out of scope for this PR and I'm almost inclined to say make the new passes a follow on PR; but that would be potentially problematic as the transpiler would error on circuits with Cliffords without the passes. So I'm fine with the shim as long as we're clear with an inline comment about how we plan to use and expand the pass to be pluggable (it'd be good if we tried to get that in for 0.22).

The only thing I was unsure of was the broadcasting piece in `QuantumCircuit.append()` and using an implicit default of `Instruction`'s broadcasting method for operations. I haven't played with the broadcasting behavior in circuits in a long time and I know @jakelishman has been thinking more about it recently. So I'm fine going forward with what you have there and we can always change it before the 0.22.0 release if needed.

alexanderivrii commented on releasenotes/notes/operation-abstract-base-class-c5efe020aa9caf46.yaml:
```diff
@@ -0,0 +1,15 @@
+    A Clifford gate is now added to a quantum circuit as an :class:`.Operation`, without first
+    synthesizing a subcircuit implementing this Clifford gate. The actual synthesis is postponed
+    to a later transpilation pass.
+  - |
```
Thanks! I have improved the release notes in [b2f972a](https://github.com/Qiskit/qiskit-terra/pull/7966/commits/b2f972ac57045bdc87108c3981d3da8bcd430620).

alexanderivrii commented on releasenotes/notes/operation-abstract-base-class-c5efe020aa9caf46.yaml:
```diff
@@ -0,0 +1,15 @@
+    to a later transpilation pass.
+  - |
+    Added a new transpiler pass :class:`.OptimizeCliffords` that collects blocks of consecutive
+    Clifford gates in a circuit, and replaces each block with a single Clifford object.
```
Thanks. I was somehow hesitant to already include `OptimizeCliffords` into preset pass managers. Maybe when more higher-level objects become available and when the pass becomes a little bit stronger?

Regarding the last point, the pass is currently restricted to Cliffords with identical qubits (viewed as ordered lists). It is likely that one can also combine Cliffords when the sets of qubits are equal (see https://github.com/Qiskit/qiskit-terra/pull/7966#discussion_r869037645), or when the sets of qubits intersect, but this requires a bit more work.

mtreinish replied:
Yeah, that's fine I think we can start with this as a standalone pass. I only mentioned that to say it was important we document how it's used since it won't be done automatically under the covers in a preset pass manager so people will be looking to the docs to understand how to use the `OptimizeCliffords` pass.

alexanderivrii commented on qiskit/quantum_info/operators/operator.py:
```diff
@@ -514,7 +514,7 @@ def _init_instruction(cls, instruction):
     def _instruction_to_matrix(cls, obj):
         """Return Operator for instruction if defined or None otherwise."""
-        if not isinstance(obj, Instruction):
+        if not isinstance(obj, Operation):
```
Oh, this is a very good point. On the one hand, we definitely want to construct an `Operator` from a quantum circuit that contain a `Clifford` or some other higher-level object:
```
cliff = Clifford(...)
qc = QuantumCircuit(5)
qc.append(cliff, [0,1,2])
op = Operator(qc)
```
On the other hand, the current code only works because `Clifford` happens to have a `to_matrix()` method, which is not required for a general `Operation`. So my "fix" above would not carry-over to future higher-level objects. Thinking aloud, the main code (see https://github.com/Qiskit/qiskit-terra/blob/2bab09c1aae84e5bf38ba52fa3a272667c1887cc/qiskit/quantum_info/operators/operator.py#L529) first checks that an `Operation` has or can have a matrix, then if it's a `Barrier`, then if it has a `definition`, and then raises an error. So ideally (as higher-level objects do not need to carry a `definition` either) we should call the `HighLevelSynthesis` pass here (or more precisely whichever routine that uses to synthesize Cliffords). But this is ugly -- as `quantum_info/Operator` would depend on transpiling routines, leading to all sorts of cyclic dependencies (run-time import and pylint-disable would fix this, but nonetheless). And I don't see another way. @mtreinish , @kdk , what do you think about this?

mtreinish replied:
What about doing:

```suggestion
        if not isinstance(obj, Operation) and hasattr(obj, "to_matrix"):
```

that way we still work with operations like clifford that have a defined matrix. We probably should include this in the operation class docs too as an optional function if we start making this an implicit part of the class (I do worry about bloating operation and having it just repeat `Instruction` though).

alexanderivrii replied:
Thanks, I see now. It should be ok that we cannot construct an `Operator` for an arbitrary quantum circuit, and in particular for circuits containing some higher-level objects (and we already can't do it for circuits containing conditional gates). Though, I completely agree, we shouldn't break this functionality for Cliffords.

It's also no clear to me if constructing `Operators` for higher-level objects based on `HighLevelSynthesis` is correct (i.e. what I was suggesting earlier) -- what if the synthesis method uses ancillas?

So I am happy to adopt the above suggestion (though, hmm, it should be `not hasattr`...). 

Agreed, I would also like to avoid bloating `Operation`. So, if I understood your suggestion correctly, I will simply include a comment in the Operation class docs.

alexanderivrii commented:
@mtreinish, thank you for the review. I have improved/expanded the description in the release notes, and have added a TODO comment in `HighLevelSynthesis`. I agree that it would be good to expand the pass to be pluggable and to do this as soon as possible, especially considering the next point.  

Even though the current code works correctly for Cliffords, as you have pointed out, there is a problem in `Operator` that will manifest for higher-level objects in the future. Basically, we need to know how to construct a matrix for any given higher-level object, so it seems that `Operator` has to be aware of `HighLevelSynthesis`. So it seems that we already need a generic `HighLevelSynthesis` plugin in order to handle `Operator` correctly. Given the size of this PR, the question is whether we really should have this plugin in this PR, or it's ok to have it in a follow-up PR, with the understanding that Operator code will not work with non-Clifford higher-level objects?

mtreinish reviewed (commented):
Thanks for making the updates. This mostly LGTM for now. The only thing I'd like to see resolved is the use of `Operation.to_matirx` in `Operator`: https://github.com/Qiskit/qiskit-terra/pull/7966/files#r941549919 I think as long as we outline that `to_matrix()` is an optional thing on operations and that the code handles the path when we have an opaque instruction without a matrix representation  that is sufficient to resolve this. Once that is fixed I think I'm fine approving this.

alexanderivrii commented:
> Thanks for making the updates. This mostly LGTM for now. The only thing I'd like to see resolved is the use of `Operation.to_matirx` in `Operator`: https://github.com/Qiskit/qiskit-terra/pull/7966/files#r941549919 I think as long as we outline that `to_matrix()` is an optional thing on operations and that the code handles the path when we have an opaque instruction without a matrix representation that is sufficient to resolve this. Once that is fixed I think I'm fine approving this.

@mtreinish, is the code comment in `Operator` (see https://github.com/Qiskit/qiskit-terra/pull/7966/commits/07733d600159a086aba001528ae36a4ef021d7a9) good enough, or would you prefer to add a comment to `Operation` docs as well. I am somehow a bit reluctant to advertise the possibility of `to_matrix()` method in `Operation` (as this goes against the idea of this API). 

Another thought is that if someone really wants to construct an `Operator` for a quantum circuit `qc` with higher-level objects, then he/she can just call `qc.decompose()` the right number of times, so it's not really a must-have functionality.

mtreinish commented:
I was thinking of having it as part of the documentation but I guess a comment is fine too if we don't want to advertise it. But I think you're right maybe we should just drop this piece as it really isn't a logical part of the operation class and revert back to using `Instruction` in the Operator class. The only concern there is from a backwards compatibility PoV if someone did:

```
qc.append(clifford)
Operator(qc)
```

that would break when it previously worked right?  Maybe we just should special case things and do an isinstance for an Instruction or a clifford explicitly and leave a comment saying it's for backwards compatibility.

alexanderivrii commented:
> ```
> qc.append(clifford)
> Operator(qc)
> ```
> 
> that would break when it previously worked right? 

Yes, this used to work, and some people (like @ShellyGarion) depend on this functionality. So we should support this for backward compatibility.

> Maybe we just should special case things and do an isinstance for an Instruction or a clifford explicitly and leave a comment > saying it's for backwards compatibility.

This is a perfect suggestion, done in https://github.com/Qiskit/qiskit-terra/pull/7966/commits/ab213f2e5fdc829100a9a92bfac6f681d0f82645. On the one hand, we don't include things to Operation's API, on the other hand there is no under-the-table non-documented functionality. And we can always revisit things in the future as more higher-level-operations will become available.

mtreinish reviewed (approved):
This LGTM now, thanks for sticking with this and making all the necessary updates.
